diff -up lftp-4.0.9/src/ftpclass.cc.race lftp-4.0.9/src/ftpclass.cc
diff -up lftp-4.0.9/src/Http.cc.race lftp-4.0.9/src/Http.cc
--- lftp-4.0.9/src/Http.cc.race	2011-12-08 10:01:01.502483920 +0100
+++ lftp-4.0.9/src/Http.cc	2011-12-08 10:01:20.009508865 +0100
@@ -1043,10 +1043,11 @@ int Http::Do()
       res=SocketConnect(sock,&peer[peer_curr]);
       if(res==-1 && errno!=EINPROGRESS)
       {
+	 int e=errno;
 	 NextPeer();
 	 LogError(0,"connect: %s\n",strerror(errno));
 	 Disconnect();
-	 if(NotSerious(errno))
+	 if(NotSerious(e))
 	    return MOVED;
 	 goto system_error;
       }
