diff --git a/src/Http.h b/src/Http.h
index 0cae686..8b656fc 100644
--- a/src/Http.h
+++ b/src/Http.h
@@ -83,11 +83,7 @@ class Http : public NetAccess
    void MoveConnectionHere(Http *o);
    int IsConnected() const
       {
-	 if(sock==-1)
-	    return 0;
-	 if(state==CONNECTING)
-	    return 1;
-	 return 2;
+         return state!=DISCONNECTED || tunnel_state!=NO_TUNNEL;
       }
 #if USE_SSL
    Ref<lftp_ssl> ssl;
