From d812466b0a9c5db37ed0541440d6d93cc24db8e5 Mon Sep 17 00:00:00 2001
From: "Alexander V. Lukyanov" <lav@yars.free.net>
Date: Mon, 23 Jan 2012 10:42:30 +0400
Subject: [PATCH] 	* FileCopy.cc: fixed ascii mode size getting for local
 files.

---
 src/FileCopy.cc | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/FileCopy.cc b/src/FileCopy.cc
index f07a8c2..b093064 100644
--- a/src/FileCopy.cc
+++ b/src/FileCopy.cc
@@ -1561,6 +1561,10 @@ int FileCopyPeerFDStream::Get_LL(int len)
       {
 	 SetDate(st.st_mtime);
 	 SetSize(st.st_size);
+#ifndef NATIVE_CRLF
+	 if(ascii)
+	    SetSize(NO_SIZE);
+#endif
       }
    }
 
-- 
1.9.3

