diff -Naur lftp-4.0.9.old/src/FileSet.cc lftp-4.0.9/src/FileSet.cc
--- lftp-4.0.9.old/src/FileSet.cc	2013-02-19 08:25:34.349481368 +0530
+++ lftp-4.0.9/src/FileSet.cc	2013-02-19 08:30:24.417224221 +0530
@@ -926,20 +926,22 @@
    int mode1=(defined&MODE?mode:
       (filetype_c=='d'?0755:(filetype_c=='l'?0777:0644)));
 
-   char usergroup[33];
-   usergroup[0]=0;
-   if(defined&(USER|GROUP))
-      sprintf(usergroup,"%.16s%s%.16s",defined&USER?user:"?",
-		  defined&GROUP?"/":"",defined&GROUP?group:"");
+   const char *usergroup="";
+      if(defined&(USER|GROUP))
+   {
+         usergroup=xstring::format("%.16s%s%.16s",defined&USER?user:"?",
+                   defined&GROUP?"/":"",defined&GROUP?group:"");
+   }
+
 
    int w=20-strlen(usergroup);
    if(w<1)
       w=1;
-   char size_str[20];
+   char size_str[21];
    if(defined&SIZE)
-      sprintf(size_str,"%*lld",w,(long long)size);
+      snprintf(size_str,sizeof(size_str),"%*lld",w,(long long)size);
    else
-      sprintf(size_str,"%*s",w,"-");
+      snprintf(size_str,sizeof(size_str),"%*s",w,"-");
 
    const char *date_str="-";
    if(defined&DATE)
